### DISCUSSION TITLE
关于 proxy 节点时间推进的疑惑

### DISCUSSION THREAD

**Discussion by wangtuanjie** [milvus-io/milvus#23638] [Q&A and General discussion]
根据设计[文档](https://github.com/milvus-io/milvus/blob/master/docs/design_docs/20211215-milvus_timesync.md)的要求 proxy 需要实时地向 rootcoord 推进各个 msgstream 的时间，相关的实现[代码](https://github.com/milvus-io/milvus/blob/2.2.6/internal/proxy/task_scheduler.go#L227-L306)。

梳理了下 proxy.Insert 接口流程：
1. baseTaskQueue.Enqueue 通过 TSO 获得最新的 timestamp 并设置到 insertTask 对应字段上。
2. 把 insertTask 任务放入队列后并通过 dmTaskQueue.commitPChanStat 接口更新了 dmTaskQueue.pChanStatisticsInfos 字段。
3. proxy.sendChannelsTimeTickLoop 通过 channelsTimeTickerImpl.TickgetMinTsStatistics 会感知到 2 的变更并同步到 rootcoord。
4. rootcoord 在收到请求后推进时间戳并向 msgstream 发送 timetick 消息。

proxy 向 mqbroker 是逐条发送数据，当 insertTask 有多条数据时需要不少执行时间。在 4 完成时，insertTask 可能还在执行中。

不知道上面是否理解有误，希望得到反馈或讨论。

**Comment by xiaofan-luan** [comment]
> 根据设计[文档](https://github.com/milvus-io/milvus/blob/master/docs/design_docs/20211215-milvus_timesync.md)的要求 proxy 需要实时地向 rootcoord 推进各个 msgstream 的时间，相关的实现[代码](https://github.com/milvus-io/milvus/blob/2.2.6/internal/proxy/task_scheduler.go#L227-L306)。
> 
> 梳理了下 proxy.Insert 接口流程：
> 
> 1. baseTaskQueue.Enqueue 通过 TSO 获得最新的 timestamp 并设置到 insertTask 对应字段上。
> 2. 把 insertTask 任务放入队列后并通过 dmTaskQueue.commitPChanStat 接口更新了 dmTaskQueue.pChanStatisticsInfos 字段。
> 3. proxy.sendChannelsTimeTickLoop 通过 channelsTimeTickerImpl.TickgetMinTsStatistics 会感知到 2 的变更并同步到 rootcoord。
> 4. rootcoord 在收到请求后推进时间戳并向 msgstream 发送 timetick 消息。
> 
> proxy 向 mqbroker 是逐条发送数据，当 insertTask 有多条数据时需要不少执行时间。在 4 完成时，insertTask 可能还在执行中。
> 
> 不知道上面是否理解有误，希望得到反馈或讨论。

proxy有一个队列，记录了所有正在执行的任务。因此如果任务没有执行完成，proxy会向rootcoord汇报更老的时间戳，也就是tt不会推进

**Reply by wangtuanjie** [reply]
> 因此如果任务没有执行完成，proxy会向rootcoord汇报更老的时间戳


这个是指在 insertTask.Execute 执行后的 dmTaskQueue.popPChanStats 内的逻辑吗？
